File: paintinventory.client/src/utils/dewpoint.js
const B = 17.62
const C = 243.12

export function dewPoint(airTempC, humidityPct) {
  if (airTempC === '' || airTempC == null || humidityPct === '' || humidityPct == null) return null
  const t = Number(airTempC)
  const rh = Number(humidityPct)
  if (!Number.isFinite(t) || !Number.isFinite(rh) || rh <= 0 || rh > 100) return null
  const gamma = Math.log(rh / 100) + (B * t) / (C + t)
  return Math.round(((C * gamma) / (B - gamma)) * 10) / 10
}

export function belowDewMargin(substrateTempC, dewPointC, margin = 3) {
  if (substrateTempC === '' || substrateTempC == null || dewPointC === '' || dewPointC == null) return false
  return Number(substrateTempC) - Number(dewPointC) < margin
}

File: PaintInventory.Server/Services/DewPoint.cs
namespace PaintInventory.Server.Services;

public static class DewPoint
{
    private const double B = 17.62;
    private const double C = 243.12;

    public static decimal? Calculate(decimal? airTempC, decimal? humidityPct)
    {
        if (airTempC is null || humidityPct is null) return null;
        if (humidityPct <= 0m || humidityPct > 100m) return null;

        var t = (double)airTempC.Value;
        var rh = (double)humidityPct.Value;
        var gamma = Math.Log(rh / 100.0) + (B * t) / (C + t);
        var td = (C * gamma) / (B - gamma);
        return Math.Round((decimal)td, 1);
    }
}